### PR TITLE
feat: introduces Observable Log for HttpAgent

### PR DESCRIPTION
**Pull request by krpeacock**
# Description

The agent currently outputs console logs, warnings, and errors to the console.

This introduces an Observer pattern, available as a public ["log"] attribute which returns a new type of AgentLog. Developers can add listeners to hook into the logs, and a new flag (`logToConsole`) is provided, which will output these logs to the console. 

This is turned off by default.

# How Has This Been Tested?

new unit tests

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: packages/agent/src/agent/http/index.ts [hunk 1]
```diff
@@ -219,9 +226,11 @@ export class HttpAgent implements Agent {
...
       if (!location) {
         this._host = new URL('https://icp-api.io');
-        console.warn(
-          'Could not infer host from window.location, defaulting to mainnet gateway of https://icp-api.io. Please provide a host to the HttpAgent constructor to avoid this warning.',
-        );
+        this.log.notify({
+          level: 'warn',
+          message:
+            'Could not infer host from window.location, defaulting to mainnet gateway of https://icp-api.io. Please provide a host to the HttpAgent constructor to avoid this warning.',
+        });
```
Comment (ericswanson-dfinity) [inline packages/agent/src/agent/http/index.ts:233]: Suggestion to simplify the calling interface
```suggestion
        this.log.warn('Could not infer host from window.location, defaulting to mainnet gateway of https://icp-api.io. Please provide a host to the HttpAgent constructor to avoid this warning.',);
```

Comment (krpeacock) [reply packages/agent/src/agent/http/index.ts]: Done, and I've also made `log` callable by having the class extend Function

File: packages/agent/src/agent/http/index.ts [hunk 2]
```diff
@@ -728,7 +746,7 @@ export class HttpAgent implements Agent {
   public async fetchRootKey(): Promise<ArrayBuffer> {
     if (!this._rootKeyFetched) {
       // Hex-encoded version of the replica root key
-      this.rootKey = ((await this.status()) as any).root_key;
+      this.rootKey = ((await this.status()) as JsonObject & { root_key: ArrayBuffer }).root_key;
```
Comment (ericswanson-dfinity) [inline packages/agent/src/agent/http/index.ts:749]: What's going on here?  This doesn't seem related to the description of the PR.

Comment (krpeacock) [reply packages/agent/src/agent/http/index.ts:749]: ah, just cleaning up a linting warning in the file since I was touching it

**Review by ericswanson-dfinity — APPROVED** [review]
